File: backend/routes/adm.py
from fastapi import APIRouter, Depends,HTTPException
from sqlalchemy import select

#Importando tabelas:
from models.gamificacao import Conquista
from models.usuarios import Usuarios
from models.aula import Aula

#Importando SHCEMAS:
from schemas.admSchema import ConquistaSchema, DeletarConquistaSchema, UsuarioSchema, DeletarAulaSchema

#dotenv
import os
from dotenv import load_dotenv

#Biblioteca de requisição
import requests

#Instânciando roteador
adm = APIRouter(prefix="/adm",tags=["adiministração"])

#Importando dependencias
from dependences import pegar_sessao , validar_token

##Carrego o .env
load_dotenv()

#Pego informações do banco e do bucket:
SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_KEY")
SUPABASE_BUCKET = os.getenv("SUPABASE_BUCKET")
SUPABASE_BUCKET2 = os.getenv("SUPABASE_BUCKET2")



@adm.post('/adicionar_conquista')
async def adicionar_conquista(conquista : ConquistaSchema,session = Depends(pegar_sessao)):
        try:
            ##Adiciono a conquista no banco de dados:
            conquista_nova = Conquista(conquista.nome_conquista, conquista.tipo_conquista, conquista.descricao,conquista.moeda,conquista.xp, conquista.condicao_conquista)
            session.add(conquista_nova)
            session.commit()

            ##Retorno mensagem de sucesso
            return {"mensagem" : "Conquista criada com sucesso!"}

        except Exception as exception:
            ##Se não der certo eu retorno o erro, e dou rollback no banco.
            session.rollback()
            raise HTTPException(status_code=400,detail=str(exception))


@adm.get('/listar_conquista')
async def listar_conquista(session = Depends(pegar_sessao)):
    ##Retorna os valores em dicionario dentro de uma lista.
    conquistas = session.execute(
        select(
            Conquista.id_conquista,
            Conquista.nome_conquista,
            Conquista.descricao,
            Conquista.tipo_conquista,
            Conquista.condicao_conquista
        )
    ).mappings().all()
    if not conquistas :
        raise HTTPException(status_code=204,detail="Não existe conquistas no banco de dados")
    return {"conquistas" : conquistas}


@adm.delete('/deletar_conquista')
async def deletar_conquista(dados : DeletarConquistaSchema ,session = Depends(pegar_sessao)):
    conquista = session.query(Conquista).filter(Conquista.id_conquista == dados.id_conquista).first()
    if conquista is None:
       raise HTTPException(status_code=404,detail="Conquista não encontrada")
    #Deleto a conquista
    session.delete(conquista)
    session.commit()

    ##Retorno a resposta
    return {"mensagem" : "Conquista deletada"}



@adm.get('/buscar_usuarios')
async def buscar_usuarios(session = Depends(pegar_sessao)):

    quantidade_usuarios = session.query(Usuarios).count()

    usuarios = session.execute(
        select(
            Usuarios.nome_usuario,
            Usuarios.created_at,
            Usuarios.id_usuario,
            Usuarios.updated_at,
            Usuarios.ativo,
            Usuarios.email,
            Usuarios.foto,
            Usuarios.banner
        )
        .order_by(Usuarios.id_usuario)
    ).mappings().all()

    #Pego a quantidade de

    #Usuários que não verificaram o email
    usuariosNaoAutenticados = session.query(Usuarios).filter(Usuarios.email_verificado == False).count()

    #Usuários que estão Desativados/Banidos
    usuariosDesativados = session.query(Usuarios).filter(Usuarios.ativo == False).count()

    return {
        "quantidade" : quantidade_usuarios,
        "usuarios" : usuarios,
        "naoAutenticados" : usuariosNaoAutenticados,
        "desativados" : usuariosDesativados
        }

@adm.patch('/banir_usuario')
async def banir_usuario(dados : UsuarioSchema ,session = Depends(pegar_sessao)):
    usuario = session.query(Usuarios).filter(Usuarios.id_usuario == dados.id_usuario).first()
    if usuario is None:
       raise HTTPException(status_code=404,detail="Usuario não encontrado")
    if (usuario.ativo == False):
        raise HTTPException(status_code=409, detail="Usuario já está Desativado")

    if usuario.admin == True:
        raise HTTPException(status_code=403, detail="Administradores Não pode ser Desativados")

    try:
        # Desativo o Usuario
        usuario.ativo = False
        usuario.motivo_banimento = dados.motivo_banimento.upper()
        session.commit()

        return {"mensagem": "Usuario Desativado"}

    except Exception as exception:
        ##Se não der certo eu retorno o erro, e dou rollback no banco.
        session.rollback()
        raise HTTPException(status_code=400,detail=str(exception))


@adm.patch('/desbanir_usuario')
async def desbanir_usuario(dados : UsuarioSchema ,session = Depends(pegar_sessao)):
    usuario = session.query(Usuarios).filter(Usuarios.id_usuario == dados.id_usuario).first()
    if usuario is None:
       raise HTTPException(status_code=404,detail="Usuario não encontrado")
    if (usuario.ativo == True):
        raise HTTPException(status_code=409, detail="Usuario já está Ativo")
    try:
        # Desativo o Usuario
        usuario.ativo = True
        usuario.motivo_banimento = None
        session.commit()

        return {"mensagem": "Usuario Ativado"}

    except Exception as exception:
        ##Se não der certo eu retorno o erro, e dou rollback no banco.
        session.rollback()
        raise HTTPException(status_code=400, detail=str(exception))


@adm.patch('/removerFoto_usuario')
async def removerFoto_usuario(dados : UsuarioSchema ,session = Depends(pegar_sessao)):
    usuario = session.query(Usuarios).filter(Usuarios.id_usuario == dados.id_usuario).first()
    if usuario is None:
        raise HTTPException(status_code=404, detail="Usuario não encontrado")

    if (usuario.foto == 'default.png'):
        raise HTTPException(status_code=404, detail="Usuário não tem foto!")

    if usuario.admin == True:
        raise HTTPException(status_code=403, detail="Administradores Não pode ter a Foto Removida")

    try:
        ##Deleto a pasta que contem o id dele no bucket
        url_delete = f"{SUPABASE_URL}/storage/v1/object/{SUPABASE_BUCKET}"
        resposta = requests.delete(
            url_delete,
            headers={
                "Authorization": f"Bearer {SUPABASE_KEY}",
                "apikey": SUPABASE_KEY,
                "Content-Type": "application/json"
            },
            json={
                "prefixes": [usuario.foto]
            }
        )
        if (resposta.status_code > 199 and resposta.status_code < 300):
            usuario.foto = 'default.png'
            session.commit()
            return {"mensagem": "Foto Removida"}
        else:
            raise HTTPException(status_code=400, detail=resposta.text)

    except Exception as exception:
        ##Se não der certo eu retorno o erro, e dou rollback no banco.
        session.rollback()
        raise HTTPException(status_code=400, detail=str(exception))


@adm.patch('/removerBanner_usuario')
async def removerBanner_usuario(dados : UsuarioSchema ,session = Depends(pegar_sessao)):
    usuario = session.query(Usuarios).filter(Usuarios.id_usuario == dados.id_usuario).first()
    if usuario is None:
        raise HTTPException(status_code=404, detail="Usuario não encontrado")

    if(usuario.banner == 'default.png'):
        raise HTTPException(status_code=404,detail="Usuário não tem banner!")

    if usuario.admin == True:
        raise HTTPException(status_code=403, detail="Administradores Não pode ter o Banner Removido")

    try:
        url_delete = f"{SUPABASE_URL}/storage/v1/object/{SUPABASE_BUCKET2}"
        resposta = requests.delete(
            url_delete,
            headers={
                "Authorization": f"Bearer {SUPABASE_KEY}",
                "apikey": SUPABASE_KEY,
                "Content-Type": "application/json"
            },
            json={
                "prefixes": [usuario.banner]
            }
        )
        if (resposta.status_code > 199 and resposta.status_code < 300):
            # Se der certo a requisição:

            usuario.banner = 'default.png'
            session.commit()
            return {"mensagem": 'Banner Removido'}
        else:
            raise HTTPException(status_code=400, detail=resposta.text)

    except Exception as exception:
        ##Se não der certo eu retorno o erro, e dou rollback no banco.
        session.rollback()
        raise HTTPException(status_code=400, detail=str(exception))

@adm.patch('/redefinir_nome')
async def redefinir_nome(dados : UsuarioSchema,session = Depends(pegar_sessao)):
    usuario = session.query(Usuarios).filter(Usuarios.id_usuario == dados.id_usuario).first()
    if usuario is None:
        raise HTTPException(status_code=404, detail="Usuario não encontrado")

    if(usuario.nome_usuario == "Usuário"):
        raise HTTPException(status_code=403, detail="Esse Nome não pode ser Redefinido")

    if usuario.admin == True:
        raise HTTPException(status_code=403, detail="Administradores Não pode ter o Nome Redefinido")

    try:
        usuario.nome_usuario = "Usuário"
        session.commit()

        return {"mensagem" : "Nome Redefinido"}
    except Exception as exception:
        ##Se não der certo eu retorno o erro, e dou rollback no banco.
        session.rollback()
        raise HTTPException(status_code=400, detail=str(exception))

@adm.get('/buscar_aulas')
async def buscar_aulas(session = Depends(pegar_sessao)):
    aulas = session.execute(
        select(Aula.id_aula,Aula.titulo,Aula.tipo, Aula.modulo,Aula.xp_bonus,Aula.moeda_bonus)
    ).mappings().all()
    if not aulas:
        raise HTTPException(status_code=404, detail="Nenhuma Aula encontrada!")

    return {"aulas" : aulas}

@adm.delete('/deletar_aula')
async def deletar_conquista(dados : DeletarAulaSchema ,session = Depends(pegar_sessao)):
    aula = session.query(Aula).filter(Aula.id_aula == dados.id_aula).first()
    if aula is None:
       raise HTTPException(status_code=404,detail="Aula não encontrada")
    #Deleto a conquista
    session.delete(aula)
    session.commit()

    ##Retorno a resposta
    return {"mensagem" : "Aula Deletada"}

File: frontend/Site/CrashWare/funcoes/adm.js
import { sleep } from "./functions";

export class Adm
{
    constructor(setPopup = null , Navegacao = null)
    {
        this.setPopup = setPopup;
        this.Navegacao = Navegacao;
    }

    async adicionar_conquista(nomeConquista,opcao,descricaoConquista,moedas,xp,condicao,setPopup)
    {
        setPopup({
                    tipo: 'aviso',
                    titulo: 'Conquista',
                    mensagem: 'Enviando Informações...'
                });

        await sleep(2000);

        

        try
        {
            const response = await fetch("https://api-crashware.onrender.com/adm/adicionar_conquista",
            {   
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({
                        nome_conquista: nomeConquista,
                        tipo_conquista: opcao,
                        descricao : descricaoConquista,
                        moeda: moedas,
                        xp: xp,
                        condicao_conquista: condicao
                    })
            })

            if (response.ok)
            {
                const resposta = await response.json()
                
                setPopup({
                    tipo: 'sucesso',
                    titulo: 'Conquista',
                    mensagem: resposta.mensagem
                });

            }else
            {
                const erro = await response.json()

                  setPopup({
                    tipo: 'erro',
                    titulo: 'Erro',
                    mensagem: erro.detail
                });
            }   
        
        }catch(error) 
        {
             setPopup({
                    tipo: 'erro',
                    titulo: 'Erro De Conexão',
                    mensagem: 'Tente novamente mais tarde...'
                });
            //Erro de conexão
            console.log("Erro:", error);
            

        }//catch

    }//add_conquista


    async listar_conquista(setPopup)
    {
         setPopup({
                    tipo: 'aviso',
                    titulo: 'Conquistas',
                    mensagem: 'Listando conquistas...'
                });

        try
        {
            const response = await fetch("https://api-crashware.onrender.com/adm/listar_conquista",
                 {
                    method: "GET"
                 });

        
            if (response.status == 204)
            {
                //Se não existir conquistas...

                //Pego o erro
                const erro = await response.json();

                setPopup({
                    tipo: 'erro',
                    titulo: 'Conquistas',
                    mensagem: erro.detail
                });

                 return;

            }else
            {
                //Se requisição der certo
                const resposta = await response.json();

                //Pego as conquistas
                const conquistas = await resposta.conquistas;

                //Guardo as conquistas no LocalStorage
                localStorage.setItem("conquistas", JSON.stringify(conquistas));

            }


        }catch(error) 
        {
             setPopup({
                    tipo: 'erro',
                    titulo: 'Erro De Conexão',
                    mensagem: 'Erro ao listar'
                });
            //Erro de conexão
            console.log("Erro:", error);
            

        }//catch
            
    }//Listar Conquistas

    async deletar_conquista(id_conquista,setPopup)
    {
        setPopup({
                    tipo: 'aviso',
                    titulo: 'Conquistas',
                    mensagem: 'Deletando conquista...'
                });

        sleep(1000)
        try
        {
            const response = await fetch("https://api-crashware.onrender.com/adm/deletar_conquista",
                 {
                    method: "DELETE",
                    headers: { "Content-Type": "application/json" },
                    body: JSON.stringify({
                        id_conquista : id_conquista
                        })
                 });


            if(response.ok)
            {
                //Exibo resposta da API
                const resposta = await response.json();
                setPopup({
                    tipo: 'sucesso',
                    titulo: 'Conquista',
                    mensagem: resposta.mensagem
                });

            }else
            {
                const erro = await response.json();

                //Exibo o erro
                
                setPopup({
                    tipo: 'erro',
                    titulo: 'Conquista',
                    mensagem: erro.detail
                });

               
            }

        }catch(error) 
        {
             setPopup({
                    tipo: 'erro',
                    titulo: 'Erro De Conexão',
                    mensagem: 'Erro ao deletar'
                });

            //Erro de conexão
            console.log("Erro:", error);

            sleep(1000)
            return;
        
        }//catch
    }//Deletar conquista



    //Total de usuários

    async carregar_usuarios(setPopup,setTotal = null)
    {
        try
        {
            const response = await fetch("https://api-crashware.onrender.com/adm/buscar_usuarios",
                {
                    method: "GET"
                });


            if(response.ok)
            {
                //Exibo resposta da API
                const resposta = await response.json();
                
                const total = resposta.quantidade;

                const usuariosNaoAutenticados = resposta.naoAutenticados;

                const desativados = resposta.desativados;

                if (setTotal  != null)
                {
                    setTotal(total)
                }else
                {
                    //Pego as informações do usuarios
                    const usuarios = resposta.usuarios;

                    //Guardo os usuarios no LocalStorage
                    localStorage.setItem("usuarios", JSON.stringify(usuarios));
                }
                


            }else
            {
                const erro = await response.json();

                //Exibo o erro
                
                setPopup({
                    tipo: 'erro',
                    titulo: 'Erro Inesperado',
                    mensagem: "Tente Novamente Mais Tarde..."
                });

               console.log("Erro ao carregar  Usuários")
            }

        }catch(error) 
        {
            //Erro de Internet OU na Requisição
            setPopup({
                tipo: 'erro',
                titulo: 'Usuarios',
                mensagem: 'Não foi possível conectar ao servidor.'
            });

            console.log("Erro Ao Tentar Carregar Usuários" + error)

        
        }//catch
    }//Carregar Usuários

    async desativar_usuario(id_usuario,setPopup)
    {
        setPopup({
                    tipo: 'aviso',
                    titulo: 'Usuário',
                    mensagem: 'Desativando Usuário....'
                });

        await sleep(1000)
        try
        {
             const response = await fetch("https://api-crashware.onrender.com/adm/banir_usuario",
                {
                    method: "PATCH",
                    headers:{  "Content-Type": "application/json" },
                     body: JSON.stringify({
                        id_usuario : id_usuario
                    })

                });


            if(response.status == 409)
            {

                setPopup({
                    tipo: 'aviso',
                    titulo: 'Usuário',
                    mensagem: 'Usuário já está Desativado'
                });

                return;

            }
            if(response.ok)
            {
                //Exibo resposta da API
                const resposta = await response.json();
                
                setPopup({
                    tipo: 'sucesso',
                    titulo: 'Usuário',
                    mensagem: resposta.mensagem
                });
                
            }else
            {
                const erro = await response.json();

                //Exibo o erro
                
                setPopup({
                    tipo: 'erro',
                    titulo: 'Usuário',
                    mensagem: erro.detail
                });

               console.log("Erro ao Desativar  Usuários" + erro.detail)

               return 403
            }
                
        }catch(error) 
        {
            //Erro de Internet OU na Requisição
            setPopup({
                tipo: 'erro',
                titulo: 'Erro ao Tentar Desativar Usuário',
                mensagem: 'Não foi possível conectar ao servidor.'
            });

            console.log("Erro Ao Tentar Banir Usuário" + error)

    
        }
        
    }//Desativar Usuario

    async ativar_usuario(id_usuario,setPopup)
    {
         setPopup({
                    tipo: 'aviso',
                    titulo: 'Usuário',
                    mensagem: 'Ativando Usuário....'
                });

        await sleep(1000)
        try
        {
             const response = await fetch("https://api-crashware.onrender.com/adm/desbanir_usuario",
                {
                    method: "PATCH",
                    headers:{  "Content-Type": "application/json" },
                     body: JSON.stringify({
                        id_usuario : id_usuario
                    })

                });


            if(response.status == 409)
            {

                setPopup({
                    tipo: 'aviso',
                    titulo: 'Usuário',
                    mensagem: 'Usuário já está Ativado'
                });

                return;

            }
            if(response.ok)
            {
                //Exibo resposta da API
                const resposta = await response.json();
                
                setPopup({
                    tipo: 'sucesso',
                    titulo: 'Usuário',
                    mensagem: resposta.mensagem
                });
                
            }else
            {
                const erro = await response.json();

                //Exibo o erro
                
                setPopup({
                    tipo: 'erro',
                    titulo: 'Usuário',
                    mensagem: erro.detail
                });

               console.log("Erro ao Ativar  Usuário" + erro.detail)
            }
                
        }catch(error) 
        {
            //Erro de Internet OU na Requisição
            setPopup({
                tipo: 'erro',
                titulo: 'Erro ao Tentar Ativar Usuário',
                mensagem: 'Não foi possível conectar ao servidor.'
            });

            console.log("Erro Ao Tentar Desbanir Usuário" + error)

    
        }

    }//Ativar Usuário
    

    async removerFoto_usuario(id_usuario,setPopup)
    {
         setPopup({
                    tipo: 'aviso',
                    titulo: 'Foto',
                    mensagem: 'Removendo Foto...'
                });

        await sleep(1000)

        try
        {
             const response = await fetch("https://api-crashware.onrender.com/adm/removerFoto_usuario",
                {
                    method: "PATCH",
                    headers:{  "Content-Type": "application/json" },
                     body: JSON.stringify({
                        id_usuario : id_usuario
                    })

                });

            if(response.ok)
            {
                //Exibo resposta da API
                const resposta = await response.json();
                
                setPopup({
                    tipo: 'sucesso',
                    titulo: 'Foto',
                    mensagem: resposta.mensagem
                });
                
            }else
            {
                const erro = await response.json();

                //Exibo o erro
                
                setPopup({
                    tipo: 'erro',
                    titulo: 'Foto',
                    mensagem: erro.detail
                });

               console.log("Erro ao Remover foto do Usuário" + erro.detail)
            }
                
        }catch(error) 
        {
            //Erro de Internet OU na Requisição
            setPopup({
                tipo: 'erro',
                titulo: 'Erro De Conexão',
                mensagem: 'Não foi possível conectar ao servidor.'
            });

            console.log("Erro Ao Tentar Remover Foto do Usuario" + error)

    
        }

    }//Remover Foto do Usuario

    async removerBanner_usuario(id_usuario,setPopup)
    {
         setPopup({
                    tipo: 'aviso',
                    titulo: 'Banner',
                    mensagem: 'Removendo Banner...'
                });

        await sleep(1000)

        try
        {
             const response = await fetch("https://api-crashware.onrender.com/adm/removerBanner_usuario",
                {
                    method: "PATCH",
                    headers:{  "Content-Type": "application/json" },
                     body: JSON.stringify({
                        id_usuario : id_usuario
                    })

                });

            if(response.ok)
            {
                //Exibo resposta da API
                const resposta = await response.json();
                
                setPopup({
                    tipo: 'sucesso',
                    titulo: 'Banner',
                    mensagem: resposta.mensagem
                });
                
            }else
            {
                const erro = await response.json();

                //Exibo o erro
                
                setPopup({
                    tipo: 'erro',
                    titulo: 'Banner',
                    mensagem: erro.detail
                });

               console.log("Erro ao Remover banner do Usuário" + erro.detail)
            }
                
        }catch(error) 
        {
            //Erro de Internet OU na Requisição
            setPopup({
                tipo: 'erro',
                titulo: 'Erro De Conexão',
                mensagem: 'Não foi possível conectar ao servidor.'
            });

            console.log("Erro Ao Tentar Remover Banner do Usuario" + error)

    
        }

    }//Remover Banner do Usuario

    async redefinir_nome(id_usuario,setPopup,setDados)
    {
         setPopup({
                    tipo: 'aviso',
                    titulo: 'Nome',
                    mensagem: 'Redefinindo Nome...'
                });

        await sleep(1000)

        try
        {
             const response = await fetch("https://api-crashware.onrender.com/adm/redefinir_nome",
                {
                    method: "PATCH",
                    headers:{  "Content-Type": "application/json" },
                     body: JSON.stringify({
                        id_usuario : id_usuario
                    })

                });

            if(response.ok)
            {
                //Exibo resposta da API
                const resposta = await response.json();

                setPopup({
                    tipo: 'sucesso',
                    titulo: 'Nome',
                    mensagem: resposta.mensagem
                });

                
            }else
            {
                const erro = await response.json();

                //Exibo o erro
                
                setPopup({
                    tipo: 'erro',
                    titulo: 'Nome',
                    mensagem: erro.detail
                });

               console.log("Erro ao Redefinir Nome" + erro.detail)

               return 403
            }
                
        }catch(error) 
        {
            //Erro de Internet OU na Requisição
            setPopup({
                tipo: 'erro',
                titulo: 'Erro De Conexão',
                mensagem: 'Não foi possível conectar ao servidor.'
            });

            console.log("Erro Ao Tentar Rdefinir Nome" + error)

    
        }

    }//Redefinir Nome
    
}//ADM